internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -392,6 +403,78 @@ func (r *BMACReconciler) reconcileBMH(ctx context.Context, bmh *bmh_v1alpha1.Bar
+// to an existing spoke cluster, the following things need to be done:
+// - Check agent to see if day2 flow needs to be triggered.
+// - Get the kubeconfig client from secret data
+// - Creates a new Machine
+// - Create BMH with externallyProvisioned set to true and set the newly created machine as ConsumerRef
+// BMH_HARDWARE_DETAILS_ANNOTATION is needed for auto approval of the CSR.
+func (r *BMACReconciler) reconcileSpokeBMH(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost, agent *adiiov1alpha1.Agent) reconcileResult {
+	// Only worker role is supported for day2 operation
+	if agent.Spec.Role != models.HostRoleWorker || agent.Spec.ClusterDeploymentName == nil {
+		r.Log.Debugf("Skipping spoke BareMetalHost reconcile for  agent %s/%s, role %s and clusterDeployment %s.", agent.Namespace, agent.Name, agent.Spec.Role, agent.Spec.ClusterDeploymentName)
+		return reconcileComplete{}
+	}
+
+	cdKey := types.NamespacedName{
+		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+		Name:      agent.Spec.ClusterDeploymentName.Name,
+	}
+	clusterDeployment := &hivev1.ClusterDeployment{}
+	if err := r.Get(ctx, cdKey, clusterDeployment); err != nil {
+		r.Log.WithError(err).Errorf("failed to get clusterDeployment resource %s/%s", cdKey.Namespace, cdKey.Name)
+		return reconcileError{err}
+	}
+
+	// If the cluster is not installed yet, we can't get kubeconfig for the cluster yet.
+	if !clusterDeployment.Spec.Installed {
+		r.Log.Infof("ClusterDeployment %s/%s for agent %s/%s is not installed yet", clusterDeployment.Namespace, clusterDeployment.Name, agent.Namespace, agent.Name)
+		// If cluster is not Installed, wait until a reconcile is trigged by a watch event
+		return reconcileComplete{}
+	}
+
+	// Secret contains kubeconfig for the spoke cluster
+	secret := &corev1.Secret{}
+	name := fmt.Sprintf(adminKubeConfigStringTemplate, clusterDeployment.Name)
+	err := r.Get(ctx, types.NamespacedName{Namespace: clusterDeployment.Namespace, Name: name}, secret)
+	if err != nil && errors.IsNotFound(err) {
+		r.Log.WithError(err).Errorf("failed to get secret resource %s/%s", clusterDeployment.Namespace, name)
+		// If secret is not found, wait until a reconcile is trigged by a watch event
+		return reconcileComplete{}
+	} else if err != nil {
+		return reconcileError{err}
+	}
+
+	spokeClient, err := getSpokeClient(secret)
+	if err != nil {
+		r.Log.WithError(err).Errorf("failed to create spoke kubeclient")
+		return reconcileError{err}
+	}
+
+	machine, err := r.ensureSpokeMachine(ctx, spokeClient, bmh, clusterDeployment)
+	if err != nil {
+		r.Log.WithError(err).Errorf("failed to create or update spoke Machine")
+		return reconcileError{err}
+	}
+
+	_, err = r.ensureSpokeBMHSecret(ctx, spokeClient, bmh)
+	if err != nil {
+		r.Log.WithError(err).Errorf("failed to create or update spoke BareMetalHost Secret")
+		return reconcileError{err}
+	}
+
+	_, err = r.ensureSpokeBMH(ctx, spokeClient, bmh, machine)
```

Comment:
The various `ensure*` functions could return a `reconcileResul` already, that way this reconcile function would not need to do a further check on the returned value from these calls. 

As it is right now, in case of failure, the reconcile step will log 2 error messages: 1 from the `ensure*` calls, and one from the reconcile function.

Reply:
Good catch. I should just be bubbling up the error in the ensure methods and let _reconcileSpokeBMH_ handle and log it. Done!

Returning result from the ensure method is so we don't have to make another get call to retrieve the same value gain.